python/cudf_polars/cudf_polars/dsl/traversal.py

```diff
@@ -0,0 +1,191 @@
+from typing import TYPE_CHECKING, Any, Generic, TypeVar, overload
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Generator, Mapping, MutableMapping
+
+    from cudf_polars.dsl import expr, ir
+    from cudf_polars.dsl.nodebase import Node
+    from cudf_polars.typing import ExprTransformer, GenericTransformer, IRTransformer
+
+
+__all__: list[str] = [
+    "traversal",
+    "reuse_if_unchanged",
+    "make_recursive",
+    "CachingVisitor",
+]
+
+
+def traversal(node: Node) -> Generator[Node, None, None]:
+    """
+    Pre-order traversal of nodes in an expression.
+
+    Parameters
+    ----------
+    node
+        Root of expression to traverse.
+
+    Yields
+    ------
+    Unique nodes in the expression, parent before child, children
+    in-order from left to right.
+    """
+    seen = {node}
+    lifo = [node]
+
+    while lifo:
+        node = lifo.pop()
+        yield node
+        for child in reversed(node.children):
+            if child not in seen:
+                seen.add(child)
+                lifo.append(child)
+
+
+# reuse_if_unchanged can either be applied to expressions, in which
+# case we need an ExprTransformer...
+@overload
+def reuse_if_unchanged(node: expr.Expr, fn: ExprTransformer) -> expr.Expr: ...
+
+
+# .. or to plan nodes (IR), in which case we need an IRTransformer
+@overload
+def reuse_if_unchanged(node: ir.IR, fn: IRTransformer) -> ir.IR: ...
+
+
+def reuse_if_unchanged(node, fn):
+    """
+    Recipe for transforming nodes that returns the old object if unchanged.
+
+    Parameters
+    ----------
+    node
+         Node to recurse on
+    fn
+         Function to transform children
+
+    Notes
+    -----
+    This can be used as a generic "base case" handler when
+    writing transforms that take nodes and produce new nodes.
+
+    Returns
+    -------
+    Existing node `e` if transformed children are unchanged, otherwise
+    reconstructed node with new children.
+    """
+    new_children = [fn(c) for c in node.children]
+    if all(new == old for new, old in zip(new_children, node.children, strict=True)):
+        return node
+    return node.reconstruct(new_children)
+
+
+U_contra = TypeVar("U_contra", bound=Hashable, contravariant=True)
+V_co = TypeVar("V_co", covariant=True)
+
+
+def make_recursive(
+    fn: Callable[[U_contra, GenericTransformer[U_contra, V_co]], V_co],
+    *,
+    state: Mapping[str, Any] | None = None,
+) -> GenericTransformer[U_contra, V_co]:
+    """
+    No-op wrapper for recursive visitors.
+
+    Facilitates using visitors that don't need caching but are written
+    in the same style.
+
+    Parameters
+    ----------
+    fn
+        Function to transform inputs to outputs. Should take as its
+        second argument a callable from input to output.
+    state
+        Arbitrary *immutable* state that should be accessible to the
+        visitor through the `state` property.
+
+    Notes
+    -----
+    All transformation functions *must* be pure.
```

Comment:
What does "pure" mean? I'm not familar with this, but are you referring to [`pure functions`](https://en.wikipedia.org/wiki/Pure_function)?

Reply:
Yes, a pure function will not mutate the input (i.e. no "side effects").

Reply:
Reworded to say "*must* be free of side-effects."